### PR TITLE
Fixes #9732: Create a relay API for remote run

### PR DESCRIPTION
**Pull request by amousset**
https://www.rudder-project.org/redmine/issues/9732

### CODE REVIEW

**Comment by amousset** [comment]
Commit modified

File: rudder-server-relay/SOURCES/relay-api/relay_api/remote_run.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+  requests.packages.urllib3.disable_warnings()
+except:
+  pass
+
+from pprint import pprint
+
+NEXTHOP = None
+RUDDER_COMMAND = "/opt/rudder/bin"
+
+''' Build a dict of node_id => nexthop_id '''
```
Comment (peckpeck) [inline rudder-server-relay/SOURCES/relay-api/relay_api/remote_run.py]: docstring are put just after the function declaration

File: rudder-server-relay/SOURCES/relay-api/relay_api/remote_run.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+  if node not in local_nodes:
+    raise ValueError("ERROR unknown node: " + str(node))
+  if "hostname" not in local_nodes[node]:
+    raise ValueError("ERROR invalid nodes file on the server for " + node)
+  return local_nodes[node]["hostname"]
+
+''' Call the remote run command locally '''
+def call_remote_run(host, uuid, classes, keep_output, asynchronous):
+  if classes:
+    classes_parameter = "-D " + classes
```
Comment (peckpeck) [inline rudder-server-relay/SOURCES/relay-api/relay_api/remote_run.py]: what is the classes format ?

File: rudder-server-relay/SOURCES/relay-api/relay_api/views.py [hunk 3]
```diff
@@ -68,6 +69,35 @@ def head_file(target_uuid, source_uuid, file_id):
...
+def remote_run_nodes():
+  try:
+    nodes = get_nodes_list(NODESLIST_FILE)
+    my_uuid = get_file_content(UUID_FILE)
+    return remote_run_generic(nodes, my_uuid, request.form['nodes'].split(','), False, request.form)
+  except Exception as e:
+    print(traceback.format_exc())
+    return format_error(e, API_DEBUGINFO)
+
+@app.route('/relay-api/remote-run/<string:node_id>', methods=['POST'])
```
Comment (peckpeck) [inline rudder-server-relay/SOURCES/relay-api/relay_api/views.py]: Shouldn't we add a /nodes in the url path ? to avoid conflicts

**Comment by amousset** [comment]
Commit modified

File: rudder-server-relay/SOURCES/relay-api/relay_api/remote_run.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  if node not in local_nodes:
+    raise ValueError("ERROR unknown node: " + str(node))
+  if "hostname" not in local_nodes[node]:
+    raise ValueError("ERROR invalid nodes file on the server for " + node)
+  return local_nodes[node]["hostname"]
+
+def call_remote_run(host, uuid, classes, keep_output, asynchronous):
+  """ Call the remote run command locally """
+  if classes:
+    classes_parameter = "-D " + classes
```
Comment (peckpeck) [inline rudder-server-relay/SOURCES/relay-api/relay_api/remote_run.py]: what is the classes format ?

Comment (amousset) [inline rudder-server-relay/SOURCES/relay-api/relay_api/remote_run.py]: comma separated list

**Comment by amousset** [comment]
Commit modified

**Pull request merged**